Avoid unnecessary overflow checking in non-Int Ranges iterating and indexing

yuyichao opened the pull request:
This basically makes the schematics of iterating through `UnitRange` identical with iterating though over a `StepRange` of the same `eltype` with step size `1`.

With this PR. (Note that this is running on my laptop instead of the faster machine I usually do benchmarks on so the overall performance is lower).

```
julia> function f{T}(n::T)
           s = zero(T)
           for i in one(T):one(T):n
               s += i
           end
           s
       end
f (generic function with 1 method)

julia> @benchmark f(Int32(1000))
================ Benchmark Results ========================
     Time per evaluation: 9.24 ns [9.12 ns, 9.36 ns]

julia> @benchmark f(Int(1000))
================ Benchmark Results ========================
     Time per evaluation: 16.87 ns [16.65 ns, 17.09 ns]
```

Fix #13864

simonster commented:
IIRC this was lifting to `Int` to deal with cases like `typemin(Int):typemax(Int)`, where `r.start == r.stop + one(T)`. Does this work with this PR? (If not, we should make sure that's actually covered by the tests...)

simonster commented:
I suspect that the issue above is the reason the tests are failing, but @IainNZ should the output really be this uninformative?

IainNZ commented:
What is uninformative? I'm seeing on https://travis-ci.org/JuliaLang/julia/jobs/89120120
```
	From worker 9:	     * ranges               Test Failed
	From worker 9:	  Expression: s == 256
	From worker 9:	   Evaluated: 0 == 256
```

and

```
	From worker 6:	     * reduce               Test Failed
	From worker 6:	  Expression: sum(collect(map(UInt8,0:255))) == 32640
	From worker 6:	   Evaluated: 0x00000000000045ca == 32640
```

simonster commented:
@IainNZ never mind. I was just looking too low in the log. Carry on.

IainNZ commented:
:+1: I was scared for a minute then

yuyichao commented:
> this was lifting to `Int` to deal with cases like `typemin(Int):typemax(Int)`

Lifting to `Int` doesn't help with this... It only help with whatever smaller than `Int` (And so `typemin(Int32):typemax(Int32)` will only work on 64bit and `typemin(Int64):typemax(Int64)` doesn't work anywhere...)

Also, this still an issue for `StepRange` because we don't have similar promotion for it. (e.g. `collect(typemin(Int8):one(Int8):typemax(Int8))`).

If we want to make sure `typemin(T):typemax(T)` that currently works still work I can fix the performance issue by just having specialization of `next()` for the builtin integer types....

simonster commented:
Sorry, I meant `typemin(Int8):typemax(Int8)`. I think specializing `next` for the built-in integer types to reinterpret instead of converting is a good idea. And if there's a way to make `typemin(Int8):one(Int8):typemax(Int8)` work properly, that would be great too.

yuyichao commented:
OK, I add specialization to skip the overflow check in `next` and `unsafe_getindex` for `(U)Int8` to `(U)Int32`. (only `next` is necessary for the original issue but I also happend to notice the same issue in `getindex`....)

The `StepRange` still have the same issue. It shouldn't be too hard to fix and I'll have a look later.

yuyichao commented:
> I was just looking too low in the log. Carry on.

I also find it a little hard to find the actual error in CI ..... Mainly because the summary printing is followed by useless backtraces and the inline `Test Failed` just looks like a very innocent output from the test (even more so than the warnings from certain tests....).....

yuyichao committed:
* Only promote to `Int` when iterating over StepRange and UnitRange for builtin
  integers that are smaller than Int.
* Avoid unnecessary overflow checking in non-Int Ranges iterating and indexing

yuyichao commented:
Since we are specializing on smallint anyway, I changed it so that the promotion to Int and the truncate cast are both only done for small int. This should also make the generic version a little more generic.

I added the test for `typemin(UInt8):one(UInt8):typemax(UInt8)`. Not sure how to test the original issue this is trying to fix (except letting @nanosoldier monitering the performance...)

tkelman commented on base/range.jl:
```diff
@@ -387,22 +387,36 @@ done(r::LinSpace, i::Int) = length(r) < i
+                Union{Int8,UInt8,Int16,UInt16})
+    global start
+    global next
+    start{T<:smallint}(r::StepRange{T}) = convert(Int, r.start)
```
seems like this would look a little better as `Base.start` instead of `global start` ?

yuyichao replied:
This style is copied from similar definitions for `length` above. I think `Base.start` might not work because this is also part of `Core.Inference`.

tkelman replied:
ah, little messy but oh well

IainNZ commented:
@yuyichao 
> I also find it a little hard to find the actual error in CI ..... Mainly because the summary printing is followed by useless backtraces and the inline Test Failed just looks like a very innocent output from the test (even more so than the warnings from certain tests....).....

What summary printing? I'm not sure I understand, nothing has changed with base Julia test output apart from slightly nicer formatting

tkelman commented:
tests now continue past failures but i think that was technically a separate pr that changed that, not directly the new test system

IainNZ commented:
Yeah that definitely wouldn't be the test system. I missed that change, I thought we wanted fail-fast?

yuyichao commented:
What Tony said, plus the failure printed at the end of the test run includes some unrelated backtrace making it harder to spot the name of the failing test (which is printed but just hard to find).

IainNZ commented:
I had no involvement in that, do you know the PR that changed it?

tkelman commented:
Same PR that added memory consumption printing, I think it was by Amit.

yuyichao commented:
Any additional comments about this PR?

yuyichao commented:
Will merge during the weekend if there's no other issue.
